### ISSUE TITLE
Adding Active Start and End Dates to Tock Projects

### ISSUE THREAD

**Issue by joshuabailes** [18F/tock#352]
The Agreements Team would like to add a feature to Tock. This would allow the Agreements team to set a start and end date for a tock project being active. This would automatically turn the project on or off depending on the dates, and record the dates as part of the tock data downloaded. 

**Acceptance Criteria:**

``` gherkin
Feature: As a Tock Admin user, I want to mark some projects as temporary, so the system should allow me to specify start and end dates for those projects.

Scenario: Admin can specify project start and end dates for new projects
Given I am a logged in admin user
When I am creating a new project
Then I should see inputs labeled "Project Start Date" and "Project End Date"
And they should have no value by default
And the inputs should be such that I am unable to specify an invalid date format (use a date picker or dropdowns for MM/DD/YYYY)
And I should be able to save a dates I specify

Scenario: Admin can update start and end dates for existing projects
Given I am a logged in admin user
And I am editing an existing project
And that project has a start date or an end date
Then I should see the project start date in the “Project Start Date” field
And I should see the project end date in the “Project End Date” field
And I should be able to update either date and save my changes

Scenario: Admin sees project start date and project end date in data export
Given I am a logged in admin user
And I there is a project that has a project start date or project end date
When I download data for that project
Then the project start date or project end date should be included in the downloaded data

Feature: As an 18F employee recording my time, I don’t want to mistakenly report time against a project that has not yet started, or has already ended, so the system should prevent me from doing that.

Scenario: 18F employee cannot report time against a project after it’s end date
Given I am a logged in 18F employee
And the current reporting period is ENTIRELY after the project end date 
When I search for projects to report time against
Then I should NOT see that project

Scenario: 18F employee cannot report time against a project before its start date
Given I am a logged in 18F employee
And the current reporting period is ENTIRELY before the project start date
When I search for projects to report time against
Then I should NOT see that project

Feature: As a consumer of the Tock API, I want to show important information about projects (including start and end dates), so if this data is available in the system, it should be returned in the API response.

Scenario: API returns project start and end dates
Given that an API request is made for data on a certain project
And the requestor is authorized to access the data
And the project has a start date and or an end date
When the system sends a response
Then the response should contain the project start and end dates
```

**Comment by joshuabailes** [comment]
@batemapf @martyspence See above

**Comment by adelevie** [comment]
@joshuabailes seems quite feasible for micro-purchase to me. cc @mtorres253.

**Comment by joshuabailes** [comment]
Great! What do you need to know from me to move it forward?

Joshua Bailes
18F Acquisitions/Agreements Lead
Joshua.Bailes@gsa.gov
202-550-6659
18F.gsa.gov <https://18f.gsa.gov/>

**Comment by joshuabailes** [comment]
@adelevie Checking in on this.

**Comment by adelevie** [comment]
@joshuabailes nothing for you at this point. We'll work on writing some objective acceptance criteria and fit this into the next batch.

**Comment by mtorres253** [comment]
Hi @joshuabailes! Before I take a stab at writing the acceptance criteria I have a few questions:

1. Are start and end dates required for every project?
2. It's best practice to store the date in UTC, but I recommend displaying the dates in EST/EDT is that OK?
3. What do you mean by turning the project "on" and "off"? What does that mean to a user (18f Employee as well as Admin user)?
4. Is there some way I can poke around as an admin (on staging or something) so that I can see what functionality is there now?

**Comment by joshuabailes** [comment]
@mtorres253 
1. Are start and end dates required for every project?
No. Some items, like Out of Office, are permanent. So, perhaps a way to set as either a date or an ongoing project.

2. It's best practice to store the date in UTC, but I recommend displaying the dates in EST/EDT is that OK?
This works. Does this mean there will be an hour entry as well? That seems unnecessary. 

3. What do you mean by turning the project "on" and "off"? What does that mean to a user (18f Employee as well as Admin user)?
This should be an admin only permission. Basically, I want the tock project to auto close on the date the agreement expires as entered into tock. Does that help?

4. Is there some way I can poke around as an admin (on staging or something) so that I can see what functionality is there now?
I don't know. There is no functionality related to dates now.

**Comment by joshuabailes** [comment]
👍 to this @mtorres253

**Comment by scottt2** [comment]
RE:
> 1. Are start and end dates required for every project?
No. Some items, like Out of Office, are permanent. So, perhaps a way to set as either a date or an ongoing project.

@joshuabailes , would a `null` end date suffice here? That way we can just use the date inputs to express intent. We'd just need to relax any validation on the end date field (keep it on the start, though).

**Comment by joshuabailes** [comment]
I want the end date field to control when it is populated. If not populated
or made "null" and can extend for all time, that is perfect. But I always
want project inactive if their end date is in the past.

Joshua Bailes
18F Acquisitions/Agreements Lead
Joshua.Bailes@gsa.gov
202-550-6659
18F.gsa.gov <https://18f.gsa.gov/>

**Comment by scottt2** [comment]
@joshuabailes Great! So related to that--could start date also be null?

I ask for two reasons--one, to avoid having to do a data migration and apply a start date to existing projects (retroactively applying start/end is always very prone to errors); and two, for indefinite projects, we may not need/want a start.

If I understand the krux of the feature it is just to control when a project is populated in an employee's dropdown. So we should be cool with nullable on start and end, I think. Thoughts?

Also for
> Scenario: Admin sees project start date and project end date in data export
Given I am a logged in admin user
And I there is a project that has a project start date or project end date
When I download data for that project
Then the project start date or project end date should be included in the downloaded data

It appears that there currently isn't export functionality in the admin site. Is part of this work augmenting the admin site to allow for export?

Thanks!

**Comment by scottt2** [comment]
Friendly bump, @joshuabailes 

**TL;DR**

1. Can start date also be null (allows for indefinite and avoids gnarly migration)?
2. Export doesn't seem to be in admin. Is adding that functionality assumed Feature 1, Scenario 3?

Thanks!

**Comment by scottt2** [comment]
Also, @mtorres253 & @joshuabailes -- I see that the delivery deadline is 24 hours past the bid auction end date. I just want to confirm that is the intent so I can plan to prioritize this work should I win.

Thanks, guys!
-Tyler

**Comment by joshuabailes** [comment]
I am on a bit of a fire drill. Will answer all of this ASAP. If deadline is
pushed that is fine.

Joshua Bailes
18F Acquisitions/Agreements Lead
Joshua.Bailes@gsa.gov
202-550-6659
18F.gsa.gov <https://18f.gsa.gov/>

**Comment by scottt2** [comment]
No worries, @joshuabailes . Really any time before the auction deadline would help me out.

Good luck!

**Comment by joshuabailes** [comment]
1. Can start date also be null (allows for indefinite and avoids gnarly
   migration)?

Yes.

   1. Export doesn't seem to be in admin. Is adding that functionality
   assumed Feature 1, Scenario 3?

I don't think this is contemplated. Currently, I download a csv of the
data. @mtorres253 thoughts?

Joshua Bailes
18F Acquisitions/Agreements Lead
Joshua.Bailes@gsa.gov
202-550-6659
18F.gsa.gov <https://18f.gsa.gov/>

**Comment by scottt2** [comment]
> Currently, I download a csv of the data.

@joshuabailes Could you screen shot in the admin UI where you do that for me, please? The only export I could in the codebase is this guy:

https://github.com/18F/tock/blob/68c1b176228194cd6f3bd498536da17c4cc4f352/tock/hours/views.py#L281

which is for reporting period, not for project.

Thanks!
-Tyler

**Comment by scottt2** [comment]
And this guy:

https://github.com/18F/tock/blob/68c1b176228194cd6f3bd498536da17c4cc4f352/tock/api/urls.py#L11

which appears to do the project list as csv or json.

**Comment by scottt2** [comment]
Ah, maybe I assumed there was a UI. It appears the admin api spits out the project data with the following URL format:

`.../api/projects/:project_id>.json`

Just to confirm, is that how you normally retrieve this information?

**Comment by joshuabailes** [comment]
I pull it from https://tock.18f.gov/api/timecards_bulk.csv cc @batemapf for further clarification

**Comment by scottt2** [comment]
@joshuabailes Cool! Thanks for this. Clarifies it on my end. Just a bit of work to the serializer, then.

**Comment by batemapf** [comment]
hey folks - i have some concerns over adding additional features into tock that do not get at its core purpose: to allow employees to track time easily. i'd very much prefer a standalone `agreements` module that feeds project information into tock. this shouldn't be crazy hard, just a fresh django app, little api work. i suspect we have some of the pieces lying around here already (like an authentication library).

let's talk.

cc @mtorres253 @joshuabailes @adelevie

**Comment by scottt2** [comment]
Hey @batemapf !

I know I'm not 18F (no actual skin in the game), but after having scoped this work extensively for the auction, the feature being added is relatively small and I'd argue it does get at the core purpose of easily tracking time.

From a user perspective (non-admin), this feature will filter out projects that are not applicable in the reporting period for which the user is filling out a timecard. So rather than have to search/browse a list of *all* projects as they currently do, users get only the projects that they likely care about.

**Comment by batemapf** [comment]
thanks @scottt2. i spoke with @mtorres253 @brendansudol and @joshuabailes this morning. we are currently developing another django app to maintain and surface (publicly) project-level information. we, of course, do not want to duplicate data and logic in both Tock and in the projects app. here is what we are thinking:

1. move forward with this micropurchase;
2. when we get a little further along in the development of the projects app, pull the logic and data (for this, and everything else related to projects) out of Tock and drop it in the projects app
3. concurrently, add write functionality to the Tock api and link the two apps so the project app serves up (only active) project data to Tock (and anywhere else it needs to go).

good news is lots of micropurhcase opportunities!

**Comment by scottt2** [comment]
@batemapf Right on! Totally get your perspective. And 👍 💯  on more micropurchase opportunities.

**Comment by scottt2** [comment]
Hey guys! Here is the PR: https://github.com/18F/tock/pull/374

**Comment by mtorres253** [comment]
@scottt2 We're going to extend the deadline for you to get the tests to pass. We're doing this because you submitted a pull request almost immediately, and you appear to be on a good path towards delivery. If you can fix the tests by tomorrow end of day ET, then you will still get the award.

Thanks again for participating and we look forward to continuing the relationship.

**Comment by scottt2** [comment]
@mtorres253 Thanks so much!

I'm all hands on this until it's resolved. I know the right solution will be very simple, I just haven't quite put the pieces together. If folks have thoughts on what I said above, please feel free to chime in.

**Comment by abrouilette** [comment]
@mtorres253, did this happen?  Can I close it?

**Comment by scottt2** [comment]
Hey @abrouilette! Does this mean the 18F micropurchase is still active? I've been hoping to see some new tasks come down the pipe. Just wondering if you're able to share publicly if the program is still running?

**Comment by abrouilette** [comment]
Hello @scottt2!  I will find you someone who can field this question.

**Comment by stvnrlly** [comment]
@scottt2 👋  the micro-purchase program is still running, but is not very active right now. We have some plans to make it more active, and will make sure to alert the listserv (which you can sign up for [here](http://gsa.us9.list-manage2.com/subscribe?u=6f1977de9eff4c384dc8d6527&id=e7f757afe3 if you haven't already) when that happens.
